src/python/gem5/prebuilt/viper/gpu_cache_hierarchy.py

```diff
@@ -0,0 +1,351 @@
+                is_cpu_sequencer=False,
+                deadlock_threshold=deadlock_threshold,
+            )
+
+            # SQC is shared across {cu_per_sqc} CUs.
+            cu_base = cu_per_sqc * idx
+            for cu_num in range(cu_per_sqc):
+                cu_id = cu_base + cu_num
+                compute_units[cu_id].sqc_port = sqc.sequencer.in_ports
+
+            sqc.ruby_system = self.ruby_gpu
+            sqc.TCC_select_num_bits = tcc_bits
+            sqc.clk_domain = self.clk_domain
+            sqc.recycle_latency = 10
+
+            self._controllers.append(sqc)
+
+        num_scalars = num_sqcs
+        for idx in range(num_scalars):
+            scalar = SQCCache(
+                sqc_size=self._scalar_size,
+                sqc_assoc=self._scalar_assoc,
+                network=self.ruby_gpu.network,
+                cache_line_size=self._cache_line_size,
+            )
+
+            # Scalar uses same controller as SQC, so add SQC count
+            scalar.version = idx + num_sqcs
+
+            scalar.sequencer = RubySequencer(
+                version=self.seqCount(),
+                dcache=scalar.L1cache,
+                ruby_system=self.ruby_gpu,
+                support_data_reqs=False,
+                is_cpu_sequencer=False,
+                deadlock_threshold=deadlock_threshold,
+            )
+
+            # Scalar cache is shared across {cu_per_sqc} CUs.
+            cu_base = cu_per_sqc * idx
+            for cu_num in range(cu_per_sqc):
+                cu_id = cu_base + cu_num
+                compute_units[cu_id].scalar_port = scalar.sequencer.in_ports
+
+            scalar.ruby_system = self.ruby_gpu
+            scalar.TCC_select_num_bits = tcc_bits
+            scalar.clk_domain = self.clk_domain
+            scalar.recycle_latency = 10
+
+            self._controllers.append(scalar)
+
+        # Create TCCs (GPU L2 cache)
+        for idx in range(tcc_count):
+            tcc = TCCCache(
+                tcc_size=self._tcc_size,
+                tcc_assoc=self._tcc_assoc,
+                network=self.ruby_gpu.network,
+                cache_line_size=self._cache_line_size,
+            )
+
+            tcc.version = idx
+
+            tcc.ruby_system = self.ruby_gpu
+            tcc.WB = False
+            tcc.clk_domain = self.clk_domain
+            tcc.recycle_latency = 10
+
+            self._controllers.append(tcc)
+
+        # Create DMA controllers
+        for i, port in enumerate(shader.get_gpu_dma_ports()):
+            ctrl = ViperGPUDMAController(
+                self.ruby_gpu.network, self._cache_line_size
+            )
+            ctrl.dma_sequencer = DMASequencer(version=i, in_ports=port)
+
+            ctrl.ruby_system = self.ruby_gpu
+            ctrl.dma_sequencer.ruby_system = self.ruby_gpu
+
+            self._dma_controllers.append(ctrl)
+
+        # Create GPU memories. Currently fixed to HBM2.
+        mem_type_cls = HBM_2000_4H_1x64
```

Comment:
We are about to push the updated HBM support to public (hopefully today) ... will need to figure out how that interacts with this ...